<launch>

    <arg name="cmd_src"        default="none"     doc="command source for turtlebot movement (circle or teleop)"/>
    <arg name="robot"          default="nusim"    doc="robot name (nusim, localhost, or (robot_name))"          />
    <arg name="nusim_rate"     default="500"      doc="rate at which nusim node publishes (default 500)"        />
    <arg name="frequency"      default="100"      doc="frequency at which circle node publishes"                />

    <remap from="/red/joint_states" to="/joint_states"/>
    <remap from="/red/wheel_cmd" to="/wheel_cmd"/>
    <remap from="/red/sensor_data" to="/sensor_data"/>

    <param name="odom_id" type="string" value="odom"/>
    <param name="body_id" type="string" value="blue_base_footprint"/>
    <param name="rate" type="" value="100"/>

    <rosparam file="$(find nuturtle_description)/config/diff_params.yaml" ns="nusim"/>
    <node name="circle" pkg="nuturtle_control" type="circle" if="$(eval arg('cmd_src')=='circle')">
        <param name="~frequency" value="$(arg frequency)"/>
    </node>
    <include file="$(find turtlebot3_teleop)/launch/turtlebot3_teleop_key.launch" if="$(eval arg('cmd_src')=='teleop')"/>
    

    <!-- Loading stuff for blue odometry robot in Rviz -->
    <param name="/nu_purple/robot_description" command="$(find xacro)/xacro '$(find nuturtle_description)/urdf/turtlebot3_burger.urdf.xacro' color:=nu_purple"/>
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" ns="nu_purple"/>
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" ns="nu_purple"/>
    <!-- <node pkg="tf2_ros" type="static_transform_publisher" name="world_to_odom" args="0 0 0 0 0 0 1 world odom" /> -->

    <param name="green/robot_description" command="$(find xacro)/xacro '$(find nuturtle_description)/urdf/turtlebot3_burger.urdf.xacro' color:=green"/>
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" ns="green"/>
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" ns="green"/>
    <node pkg="tf2_ros" type="static_transform_publisher" name="world_to_map" args="0 0 0 0 0 0 1 world map" />

    <!-- When launching real robot -->
    <group if="$(eval arg('robot')!='nusim')">
        <node name="turtle_interface" pkg="nuturtle_control" type="turtle_interface" machine="turtlebot"/>
        <node name="slam" pkg="nuslam" type="slam" machine="turtlebot"/>
        <node name="landmarks" pkg="nuslam" type="landmarks"/>
        <include file="$(find nuturtle_control)/launch/basic_remote.launch" pass_all_args="true"/>
        <node pkg="rviz" type="rviz" name="rviz" required="true" args="-f world -d $(find nuturtle_control)/config/nuturtle_control_2.rviz"/>
    </group>

    <!-- When launching simulation -->
    <group if="$(eval arg('robot')=='nusim')">
        <node name="nusim" pkg="nusim" type="nusim_node">
            <param name="~rate" value="$(arg nusim_rate)"/>
        </node>
        <node name="turtle_interface" pkg="nuturtle_control" type="turtle_interface"/>
        <node name="slam" pkg="nuslam" type="slam"/>
        <node name="landmarks" pkg="nuslam" type="landmarks"/>
        <rosparam file="$(find nusim)/config/basic_world.yaml"/>
        <param name="/red/robot_description" command="$(find xacro)/xacro '$(find nuturtle_description)/urdf/turtlebot3_burger.urdf.xacro' color:=red"/>
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" ns="red"/>
        <node pkg="rviz" type="rviz" name="rviz" required="true" args="-f world -d $(find nuslam)/config/landmarks.rviz"/>
    </group>

</launch>